packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.kt

```diff
@@ -0,0 +1,151 @@
+ * was a delete; if it moves forward by a character, likely to be a key press of that character.
+ *
+ * <p>IMEs can also call {@link InputConnection#beginBatchEdit()} to signify a batch of operations.
+ * One such example is committing a word currently in composing state with the press of the space
+ * key. It is IME dependent but the stock Android keyboard behavior seems to be to commit the
+ * currently composing text with {@link InputConnection#setComposingText(CharSequence, int)} and
+ * commits a space character with a separate call to {@link
+ * InputConnection#setComposingText(CharSequence, int)}. Here we chose to emit the last input of a
+ * batch edit as that tends to be the user input, but it's completely arbitrary.
+ *
+ * <p>Another function of this class is to detect backspaces when the cursor at the beginning of the
+ * {@link android.widget.EditText}, i.e no text is deleted.
+ *
+ * <p>N.B. this class is only applicable for soft keyboards behavior. For hardware keyboards {@link
+ * android.view.View#onKeyDown(int, KeyEvent)} can be overridden to obtain the keycode of the key
+ * pressed.
+ */
+public class ReactEditTextInputConnectionWrapper(
+  target: InputConnection,
+  reactContext: ReactContext,
+  private val editText: ReactEditText,
+  private val eventDispatcher: EventDispatcher
+) : InputConnectionWrapper(target, false) {
+
+  public companion object {
+    public const val NEWLINE_RAW_VALUE: String = "\n"
+    public const val BACKSPACE_KEY_VALUE: String = "Backspace"
+    public const val ENTER_KEY_VALUE: String = "Enter"
+  }
+
+  private var isBatchEdit = false
+  private var key: String? = null
+
+  override fun beginBatchEdit(): Boolean {
+    isBatchEdit = true
+    return super.beginBatchEdit()
+  }
+
+  override fun endBatchEdit(): Boolean {
+    isBatchEdit = false
+    key?.let {
+      dispatchKeyEvent(it)
+      key = null
+    }
+    return super.endBatchEdit()
+  }
+
+  override fun setComposingText(text: CharSequence?, newCursorPosition: Int): Boolean {
+    val previousSelectionStart = editText.selectionStart
+    val previousSelectionEnd = editText.selectionEnd
+
+    val consumed = super.setComposingText(text, newCursorPosition)
+
+    val currentSelectionStart = editText.selectionStart
+    val noPreviousSelection = previousSelectionStart == previousSelectionEnd
+    val cursorDidNotMove = currentSelectionStart == previousSelectionStart
+    val cursorMovedBackwardsOrAtBeginningOfInput =
+      currentSelectionStart < previousSelectionStart || currentSelectionStart <= 0
+
+    val inputKey =
+      if (cursorMovedBackwardsOrAtBeginningOfInput || (!noPreviousSelection && cursorDidNotMove)) {
+        BACKSPACE_KEY_VALUE
+      } else {
+        editText.text?.get(currentSelectionStart - 1).toString()
+      }
+
+    dispatchKeyEventOrEnqueue(inputKey)
+    return consumed
+  }
+
+  override fun commitText(text: CharSequence?, newCursorPosition: Int): Boolean {
```

Comment:
Similar here, let's try to keep it non-nullable as in the original Java file

```suggestion
  override fun commitText(text: CharSequence, newCursorPosition: Int): Boolean {
```